### PR TITLE
Ops update batch

### PR DESCRIPTION
**Pull request by AlexanderRichert-NOAA**
This PR makes a batch of package changes to support the implementation of a single operational stack on WCOSS2. Changes fall into the categories of:
- Add static-only build option
- Add PIC option
- Enable static dependencies
- Add MADIS package

spack-stack changes needed:
- fms `fpic`->`pic`

Also including commit badaaf7092b04713a24bb246eaafbb1567d32e0b from main spack to resolve CI issue (https://github.com/spack/spack/pull/37708).

### CODE REVIEW

**Comment by climbfuji** [comment]
Since this PR is changing so many spack packages, do you plan getting this into the authoritative repo first so that we don't have to deal with all the differences later on?

**Comment by AlexanderRichert-NOAA** [comment]
I'm openminded, but I'd like to first put everything in our spack in one big batch to give NCO the option of testing/using it in the next few weeks/months. Most of the package changes are minor but this way it's more of a guarantee that we can get things in quickly.

**Comment by climbfuji** [comment]
> I'm openminded, but I'd like to first put everything in our spack in one big batch to give NCO the option of testing/using it in the next few weeks/months. Most of the package changes are minor but this way it's more of a guarantee that we can get things in quickly.

Ok, sounds good!

**Review by Hang-Lei-NOAA — APPROVED** [review]

File: var/spack/repos/builtin/packages/ecflow/package.py [hunk 1]
```diff
@@ -73,23 +75,41 @@ def patch(self):
...
             self.define_from_variant("ENABLE_UI", "ui"),
             self.define_from_variant("ENABLE_GUI", "ui"),
             self.define_from_variant("ENABLE_SSL", "ssl"),
             # https://jira.ecmwf.int/browse/SUP-2641#comment-208943
             self.define_from_variant("ENABLE_STATIC_BOOST_LIBS", "static_boost"),
-            self.define("Python3_EXECUTABLE", self.spec["python"].package.command),
-            self.define("BOOST_ROOT", self.spec["boost"].prefix),
+            self.define("Python3_EXECUTABLE", spec["python"].package.command),
+            self.define("BOOST_ROOT", spec["boost"].prefix),
+            self.define("CMAKE_POSITION_INDEPENDENT_CODE", "ON"),
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/ecflow/package.py]: Should this be only if `~shared`?

File: var/spack/repos/builtin/packages/gdal/package.py [hunk 2]
```diff
@@ -711,16 +719,40 @@ def configure_args(self):
...
+            deps += ["sqlite", "libtiff", "curl", "openssl"]
+        for dep in deps:
+            if dep not in self.spec:
+                continue
+            ldflags.append(self.spec[dep].libs.search_flags)
+            libs.append(self.spec[dep].libs.link_flags)
+        if self.spec.satisfies("+iconv") and self.spec["libiconv"].satisfies("~shared"):
+            libs.append("-liconv")
+
+        if ldflags: # or libs:
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/gdal/package.py]: Should this be `if ldflags or libs:`?

File: var/spack/repos/builtin/packages/harfbuzz/package.py [hunk 3]
```diff
@@ -150,4 +150,29 @@ def configure_args(self):
...
 
+        ldflags = []
+        libs = []
+        for lib in ["bzip2", "zlib", "libpng"]:
+            spec = self.spec[lib]
+            if spec.satisfies("+shared") or spec.satisfies("libs=shared"):
+                continue
+            ldflags.append(spec.libs.ld_flags)
+            libs.append(spec.libs.link_flags)
+        cairo = self.spec["cairo"]
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/harfbuzz/package.py:161]: Seeing this logic for cairo if `~shared` repeatedly. Wondering if there is a way to bake this into the cairo package (`setup_dependent_build_environment`, `libs`) instead?

This is probably a more general question that each package should set up the dependent build environment correctly, whether it's built statically or not. But maybe that only sounds good in theory but doesn't work in practice?

File: var/spack/repos/builtin/packages/madis/package.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/madis/package.py:1]: I wonder if this addition should be its own PR. That will make it easier to cherry-pick and send to the authoritative repo.

**Review by climbfuji — COMMENTED** [review]
This is an impressive amount of work!

Do we have a sense of how likely it is that spack will accept these changes? What is our strategy sending those back up to the authoritative repo?

I have a few comments and suggestions for individual files, but nothing major. I'd like to merge this today, if possible.

File: var/spack/repos/builtin/packages/ecflow/package.py (hunk 1 above)
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/ecflow/package.py]: I just added a pic variant to make it more consistent with other packages.

File: var/spack/repos/builtin/packages/harfbuzz/package.py (hunk 3 above)
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/harfbuzz/package.py:161]: Yeah I know what you mean and it's something I've been thinking about for a while. I think I mentioned in the Spack survey a couple months back that it would be nice for there to be cleaner integration of lib/linker flags when using static builds (which creates the special need for all this `if ~shared` logic). But I'm not sure what a solution there would even look like, because each build system doesn't necessarily, for instance, use `LIBS` and `LDFLAGS` variables/arguments in the same way. For the moment all I can say is that I think this approach is broadly consistent with how things have been done in any number of other packages. In the future, if nothing else, maybe some of the logic can be simplified by having a package spec's `libs` attribute provide lib/linker flags for its dependencies and not just its own libraries based on `find_library`.

File: var/spack/repos/builtin/packages/madis/package.py (hunk 4 above)
Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/madis/package.py:1]: I think there's just two commits for madis (9dd367080cf3a45e977588355c2ce162e42f954b and 1c7889135ff466197aeae0b6329515e0e431bd97), so I don't think cherry-picking should be an issue, but if you think it's better to have a separate PR I can make one.

Comment (climbfuji) [reply var/spack/repos/builtin/packages/madis/package.py:1]: I guess it's ok. Thanks for making the changes I requested. Please update from jcsda_emc_spack_stack and we're good to go!

**Review by climbfuji — APPROVED** [review]
